### PR TITLE
Use cache to avoid reprocessing unchanged files

### PR DESCRIPTION
**Pull request by dmarkow**
Closes: #5629. Caching file transforming results on unchanged files brings rebuild times back down to pre-1.14.0 levels.

- [ ] Docs
- [ ] Tests

Testing Strategy: HMR integration test is still passing. Used it this week on my primary app and HMR seems to still work as expected, but with much quicker build times.

```bash
$ playwright test --config ./integration/playwright.config.ts -j 1 integration/hmr-test.ts

Running 4 tests using 1 worker
[chromium] › hmr-test.ts:185:1 › HMR
[webkit] › hmr-test.ts:185:1 › HMR
[edge] › hmr-test.ts:185:1 › HMR
[firefox] › hmr-test.ts:185:1 › HMR

  4 passed (33s)
```

### CODE REVIEW

**Comment by ZipBrandon** [comment]
@dmarkow Thanks for making this. I patch-packaged it onto 1.14.0 and my rebuild time is around 5.5 seconds compared to 9 or 10.

**Comment by dmarkow** [comment]
@pcattori This is the PR I mentioned on Discord. just rebased to the current dev branch and tests still passing. 700 route app was at 10-12s rebuilds before this, 2-3s after. Still not ideal, but almost an order of magnitude better. 

Uses the caching strategy outlined at https://esbuild.github.io/plugins/#caching-your-plugin to avoid re-transforming/writing files that haven't actually changed.

I tried to find a way to speed up initial builds too, but I think babel's transform function was where most of the time was spent (it's been a little while since I dug into this). We could add a file-based cache, but that adds a lot more complexity and potential for stale cache files.

**Comment by dmarkow** [comment]
Tested my patch on the released 1.16.0 today and it still works. Time savings aren't quite as good now, unfortunately -- I'm at around 4 second rebuilds now with a 12 second clean build, whereas I was getting low 2 second rebuilds on 1.15 with my patch. This may be related to the css plugins though -- I noticed a fair amount of time is spent compiling the same css multiple times which may be part of that slowdown, I'll dig in when/if I get more time.

**Comment by pcattori** [comment]
Thanks @dmarkow! With new dev server architecture done, I'm now laser-focused on bug fixes and perf improvements for compiler / dev server.

File: packages/remix-dev/compiler/js/plugins/routes_unstable.ts [hunk 1]
```diff
@@ -130,28 +132,42 @@ export function browserRouteModulesPlugin(
 
           let sourceCode = fs.readFileSync(routeFile, "utf8");
```
Comment (dennishall) [inline packages/remix-dev/compiler/js/plugins/routes_unstable.ts:133]: @dmarkow - What do you think of this idea? Maybe it's faster? Check for filechange by using `fs.statSync` -- before reading the file's contents -- and if stats.mtime or stats.size differ from cache, then file has changed? Could save the time of reading the whole file every time to see if it has changed.

Comment (dennishall) [reply packages/remix-dev/compiler/js/plugins/routes_unstable.ts:133]: Surprisingly underwhelming result. Thanks for checking it out!

**Pull request merged**